### ISSUE TITLE
Clang Static Analyzer identifies a potential fault in rsa.c

### ISSUE THREAD

**Issue by aaronmdjones** [ARMmbed/mbedtls#1002]
**(I don't know if this is a false positive)**

```
mbedtls/library/rsa.c:1493:14: warning: The left operand of '!=' is a garbage value
    if( *p++ != 0 || *p++ != MBEDTLS_RSA_SIGN )
        ~~~~ ^
1 warning generated.
```

I have uploaded the report to [1] and you can click on the hyperlink at the very top of the page to follow the control flow. You're welcome to copy it.

[1] [https://www-hosting.exbit.io/~aaron/files/report-e97d01.html](https://www-hosting.exbit.io/~aaron/files/report-e97d01.html)

**Comment by RonEld** [comment]
Hi @aaronmdjones Thank you for sharing this information.
We will investigate this issue to identify whether it is a false positive or not

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-1535

**Comment by mpg** [comment]
I believe this is a false positive, caused by the fact that clang-SA doesn't understand that `rsa_public()` will only return `0` is it successfully wrote to the output buffer. I believe clang-SA misses that because at this point it doesn't know that `mpi_read_binary()` can only return `0` or `MBEDTLS_ERR_MPI_ALLOC_FAILED`, so that at line 324 `MBEDTLS_ERR_RSA_PUBLIC_FAILED + ret` is  never `0`.

Such false positives are not uncommon with clang static analyzer on the mbed TLS code base in my experience.

**Comment by PeterHuewe** [comment]
If you run scan-build with a higher -maxloop value (e.g. 50), clang/scan-build gets it and does not show this false-positve.
Thanks for investigating though!

**Comment by aaronmdjones** [comment]
I should clarify that I'm running the latest scan-build from SVN trunk (currently `308419`) and had a `-maxloop 128` argument when reporting this issue (and the error still shows).

```
+ scan-build-5.0 -V -maxloop 128 -o scan-build/ scripts/clang-build.sh
scan-build: Using '/usr/lib/llvm-5.0/bin/clang' for static analysis
+ /usr/share/clang/scan-build-5.0/bin/../libexec/ccc-analyzer -std=c99 [...] -o foo mbedtls/library/aes.c mbedtls/library/aesni.c mbedtls/library/arc4.c [...] mbedtls/library/x509write_csr.c mbedtls/library/xtea.c src/foo.c
mbedtls/library/rsa.c:1493:14: warning: The left operand of '!=' is a garbage value
    if( *p++ != 0 || *p++ != MBEDTLS_RSA_SIGN )
        ~~~~ ^
1 warning generated.
```

**Comment by PeterHuewe** [comment]
I tried again and  had the message gone with -maxloops 300 and scan-build-3.9

```
$ scan-build -maxloop 200 make -C library/ -j24 
scan-build: Using '/usr/bin/clang-3.9' for static analysis
make: Entering directory '/fast/mbedtls-github/library'
  CC    rsa.c
rsa.c:1380:16: warning: The left operand of '>>' is a garbage value
    if( buf[0] >> ( 8 - siglen * 8 + msb ) )
        ~~~~~~ ^
rsa.c:1493:14: warning: The left operand of '!=' is a garbage value
    if( *p++ != 0 || *p++ != MBEDTLS_RSA_SIGN )
        ~~~~ ^
2 warnings generated.
  AR    libmbedcrypto.a
  RL    libmbedcrypto.a
make: Leaving directory '/fast/mbedtls-github/library'
scan-build: 2 bugs found.
scan-build: Run 'scan-view /tmp/scan-build-2017-07-29-024402-25483-1' to examine bug reports.

$ rm library/rsa.o

$ scan-build -maxloop 300 make -C library/ -j24 
scan-build: Using '/usr/bin/clang-3.9' for static analysis
make: Entering directory '/fast/mbedtls-github/library'
  CC    rsa.c
  AR    libmbedcrypto.a
  RL    libmbedcrypto.a
make: Leaving directory '/fast/mbedtls-github/library'
scan-build: Removing directory '/tmp/scan-build-2017-07-29-024428-9804-1' because it contains no reports.
scan-build: No bugs found.
```

**Comment by aaronmdjones** [comment]
The error is still showing with `-maxloop 300` on 5.0, but you are correct that 3.9 does not show it. This may be a regression in scan-build.

**Comment by aaronmdjones** [comment]
(on the other hand, it may be a real bug that 3.9 incorrectly misses under higher iteration count, I simply don't know)

**Comment by mpg** [comment]
I'm convinced this is a false positive as explained [above](https://github.com/ARMmbed/mbedtls/issues/1002#issuecomment-314713353). Also I can't reproduce this with the released version of clang 5.0. Steps taken: `scan-build-5.0 make lib` (on current development) with `clang version 5.0.0-3~16.04.1 (tags/RELEASE_500/final)`.

So I'm inclined to close this issue as "not a bug".

**Comment by mpg** [comment]
Closing this issue as "not a bug" as explained above. If you disagree, please re-open and explain why. Thanks again for sharing this information with us so we could investigate.
